hash-parser/src/location.rs

```diff
@@ -0,0 +1,66 @@
+//! Frontend-agnostic token/input locations utilities and definitions.
+//
+// All rights reserved 2021 (c) The Hash Language authors
+use crate::grammar::Rule;
+use std::fmt;
+
+/// Enum representing a location of a token within the source.
+#[derive(Debug, Eq, Clone, PartialEq)]
+pub enum Location {
+    Pos(usize),
+    Span(usize, usize),
+}
+
+/// General functions to create [Location] variants
+impl Location {
+    /// Create a 'Pos' variant by providing a single position
+    pub fn pos(pos: usize) -> Location {
+        Location::Pos(pos)
+    }
+
+    /// Create a 'Span' variant by providing a single position and the span of the input token
+    pub fn span(start: usize, end: usize) -> Location {
+        Location::Span(start, end)
+    }
+}
+
+/// Convert a [pest::error::InputLocation] into a [Location]
+impl From<pest::error::InputLocation> for Location {
+    fn from(pest: pest::error::InputLocation) -> Location {
+        match pest {
+            pest::error::InputLocation::Pos(pos) => Location::Pos(pos), // we could just use Location::Span(pos, 1) here
+            pest::error::InputLocation::Span((start, end)) => Location::Span(start, end),
+        }
+    }
+}
```

Comment:
Would it be a good idea to have all `pest`-related things be under the `pest_parser` module?

Reply:
Do you want to make multiple parsing frontends with this approach?

Reply:
Never mind, that's not possible for foreign trait implementations.

Reply:
Well I think it is generally a good idea to keep external things separate. If we ever want to make our own parser (likely) it will be very easy to do so if the usage of `pest` is isolated in the codebase.

Reply:
Yes, I think this could be moved into pest_parser or something equivalent. I think we can do this as soon as we figure out the 'ast-ification' process.